tests/update_workflow_test.go

```diff
@@ -3918,3 +3919,471 @@ func (s *integrationSuite) TestUpdateWorkflow_CompletedSpeculativeWorkflowTask_D
 		})
 	}
 }
+
+func (s *integrationSuite) TestUpdateWorkflow_StaledSpeculativeWorkflowTask_DifferentStartedId_Rejected() {
+	tv := testvars.New(s.T().Name())
+	tv = s.startWorkflow(tv)
+
+	testCtx := NewContext()
+	wtHandlerCalls := 0
+	wtHandler := func(execution *commonpb.WorkflowExecution, wt *commonpb.WorkflowType, previousStartedEventID, startedEventID int64, history *historypb.History) ([]*commandpb.Command, error) {
+		wtHandlerCalls++
+		switch wtHandlerCalls {
+		case 1:
+			// Schedule activity.
+			return []*commandpb.Command{{
+				CommandType: enumspb.COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK,
+				Attributes: &commandpb.Command_ScheduleActivityTaskCommandAttributes{ScheduleActivityTaskCommandAttributes: &commandpb.ScheduleActivityTaskCommandAttributes{
+					ActivityId:             tv.ActivityID("5"),
+					ActivityType:           tv.ActivityType(),
+					TaskQueue:              tv.TaskQueue(),
+					ScheduleToCloseTimeout: tv.InfiniteTimeout(),
+				}},
+			}}, nil
+		case 2:
+			return nil, nil
+		default:
+			s.Failf("wtHandler called too many times", "wtHandler shouldn't be called %d times", wtHandlerCalls)
+			return nil, nil
+		}
+	}
+
+	atHandler := func(execution *commonpb.WorkflowExecution, activityType *commonpb.ActivityType,
+		activityID string, input *commonpb.Payloads, taskToken []byte) (*commonpb.Payloads, bool, error) {
+		return payloads.EncodeString(tv.String("activity-result")), false, nil
+	}
+
+	poller := &TaskPoller{
```

Comment:
I personally find this TaskPoller API more confusing than helpful.
Can we use the SDK for this test or alternatively issue calls to the service directly?

Reply:
The poller is meat to be a helper and save some test code. But it is also limited. So I did have to use some of the API directly (to simulate task processing taking longer time between poll and respond).

Reply:
To use SDK here would be much harder.